File: rootfs/faasten/runtimes/python3/syscalls.py
import syscalls_pb2
import socket
import struct
import json
from contextlib import contextmanager

### helper functions ###
def recvall(sock, n):
    # Helper function to recv n bytes or return None if EOF is hit
    data = bytearray()
    while len(data) < n:
        packet = sock.recv(n - len(data))
        if not packet:
            return None
        data.extend(packet)
    return data
### end of helper functions ###

class Syscall():
    def __init__(self, sock):
        self.sock = sock

    def _send(self, obj):
        objData = obj.SerializeToString()
        try:
            self.sock.sendall(struct.pack(">I", len(objData)))
            self.sock.sendall(objData)
        except:
            while True:
                continue

    def _recv(self, obj):
        data = self.sock.recv(4, socket.MSG_WAITALL)
        res = struct.unpack(">I", data)
        objData = recvall(self.sock, res[0])

        obj.ParseFromString(objData)
        return obj

    def request(self):
        request = syscalls_pb2.Request()
        return self._recv(request)

    def respond(self, response):
        response = syscalls_pb2.Syscall(response = syscalls_pb2.Response(payload = json.dumps(response)))
        self._send(response)

    def write_key(self, key, value):
        req = syscalls_pb2.Syscall(writeKey = syscalls_pb2.WriteKey(key = key, value = value))
        self._send(req)
        response = self._recv(syscalls_pb2.WriteKeyResponse())
        return response.success

    def read_key(self, key):
        req = syscalls_pb2.Syscall(readKey = syscalls_pb2.ReadKey(key = key))
        self._send(req)
        response = self._recv(syscalls_pb2.ReadKeyResponse())
        return response.value

    ### label APIs ###
    def get_current_label(self):
        req = syscalls_pb2.Syscall(getCurrentLabel = syscalls_pb2.GetCurrentLabel())
        self._send(req)
        response = self._recv(syscalls_pb2.DcLabel())
        return response

    def taint(self, label):
        req = syscalls_pb2.Syscall(taintWithLabel = label)
        self._send(req)
        response = self._recv(syscalls_pb2.DcLabel())
        return response

    def declassify(self, secrecy: syscalls_pb2.DcComponent):
        """Declassify to the target secrecy and leave integrity untouched.
        """
        req = syscalls_pb2.Syscall(declassify = syscalls_pb2.DcComponent(value=secrecy))
        self._send(req)
        response = self._recv(syscalls_pb2.WriteKeyResponse())
        return response.success
    ### end of label APIs ###

    ### github APIs ###
    def github_rest_get(self, route):
        req = syscalls_pb2.Syscall(githubRest = syscalls_pb2.GithubRest(verb = syscalls_pb2.HttpVerb.GET, route = route, body = None))
        self._send(req)
        response= self._recv(syscalls_pb2.GithubRestResponse())
        return response

    def github_rest_post(self, route, body):
        bodyJson = json.dumps(body)
        req = syscalls_pb2.Syscall(githubRest = syscalls_pb2.GithubRest(verb = syscalls_pb2.HttpVerb.POST, route = route, body = bodyJson))
        self._send(req)
        response= self._recv(syscalls_pb2.GithubRestResponse())
        return response

    def github_rest_put(self, route, body):
        bodyJson = json.dumps(body)
        req = syscalls_pb2.Syscall(githubRest = syscalls_pb2.GithubRest(verb = syscalls_pb2.HttpVerb.PUT, route = route, body = bodyJson))
        self._send(req)
        response= self._recv(syscalls_pb2.GithubRestResponse())
        return response

    def github_rest_delete(self, route, body):
        bodyJson = json.dumps(body)
        req = syscalls_pb2.Syscall(githubRest = syscalls_pb2.GithubRest(verb = syscalls_pb2.HttpVerb.DELETE, route = route, body = bodyJson))
        self._send(req)
        response= self._recv(syscalls_pb2.GithubRestResponse())
        return response
    ### end of github APIs ###

    def invoke(self, function, payload):
        req = syscalls_pb2.Syscall(invoke = syscalls_pb2.Invoke(function = function, payload = payload))
        self._send(req)
        response= self._recv(syscalls_pb2.InvokeResponse())
        return response.success

    ### named data object syscalls ###
    def fs_read(self, path):
        """Read the file at path `path`."""
        req = syscalls_pb2.Syscall(fsRead = syscalls_pb2.FSRead(path = path))
        self._send(req)
        response = self._recv(syscalls_pb2.ReadKeyResponse())
        return response.value

    def fs_write(self, path, data):
        """Overwrite the file at path `path` with data `data`.
        The backend handler always endorse before writing.
        """
        req = syscalls_pb2.Syscall(fsWrite = syscalls_pb2.FSWrite(path = path, data = data))
        self._send(req)
        response = self._recv(syscalls_pb2.WriteKeyResponse())
        return response.success

    def fs_createdir(self, name, path, label: syscalls_pb2.DcLabel=None):
        """Create a directory `name` with label `label` in the path `path`.
        The backend handler always endorse before creating the directory.
        """
        req = syscalls_pb2.Syscall(fsCreateDir = syscalls_pb2.FSCreateDir(
            basePath = path, name = name, label = label))
        self._send(req)
        response = self._recv(syscalls_pb2.WriteKeyResponse())
        return response.success

    def fs_createfile(self, name, path, label: syscalls_pb2.DcLabel=None):
        """Create a file `name` with label `label` in the directory `path`.
        The backend handler always endorse before creating the file.
        """
        req = syscalls_pb2.Syscall(fsCreateFile = syscalls_pb2.FSCreateFile(
            basePath = path, name = name, label = label))
        self._send(req)
        response = self._recv(syscalls_pb2.WriteKeyResponse())
        return response.success
    ### end of named data object syscalls ###

    ### unnamed data object syscalls #
    @contextmanager
    def create_unnamed(self, size: int = None):
        """Create a nameless data object.

        The implementation uses the content-addressed blob store.

        Yield:
            An instance of class NewBlob
        """
        req = syscalls_pb2.Syscall(createBlob=syscalls_pb2.BlobCreate(size=size))
        self._send(req)
        response = self._recv(syscalls_pb2.BlobResponse())
        if response.success:
            fd = response.fd
            yield NewBlob(fd, self)
            syscalls_pb2.Syscall(closeBlob=syscalls_pb2.BlobClose(fd=fd))
            self._send(req)
            response = self._recv(syscalls_pb2.BlobResponse())
        else:
            raise CreateUnnamedError

    @contextmanager
    def open_unnamed(self, name):
        """Open an existing nameless data object read-only.

        The implementation uses the content-addressed blob store.

        Yield:
            An instance of class Blob
        """
        req = syscalls_pb2.Syscall(openBlob=syscalls_pb2.BlobOpen(name=name))
        self._send(req)
        response = self._recv(syscalls_pb2.BlobResponse())
        fd = response.fd
        yield Blob(fd, self)
        req = syscalls_pb2.Syscall(closeBlob=syscalls_pb2.BlobClose(fd=fd))
        self._send(req)
        response = self._recv(syscalls_pb2.BlobResponse())
    ### end of unnamed object syscalls ###

class NewBlob():
    def __init__(self, fd, syscall):
        self.fd = fd
        self.syscall = syscall

    def write(self, data):
        req = syscalls_pb2.Syscall(writeBlob=syscalls_pb2.BlobWrite(fd=self.fd, data=data))
        self.syscall._send(req)
        response = self.syscall._recv(syscalls_pb2.BlobResponse())
        return response.success

    def finalize(self, data):
        req = syscalls_pb2.Syscall(finalizeBlob=syscalls_pb2.BlobFinalize(fd=self.fd, data=data))
        self.syscall._send(req)
        response = self.syscall._recv(syscalls_pb2.BlobResponse())
        return response.data.decode("utf-8")

class Blob():
    def __init__(self, fd, syscall):
        self.fd = fd
        self.syscall = syscall

    def _blob_read(self, offset=None, length=None):
        req = syscalls_pb2.Syscall(readBlob=syscalls_pb2.BlobRead(fd=self.fd, offset=offset, length=length))
        self.syscall._send(req)
        response = self.syscall._recv(syscalls_pb2.BlobResponse())
        if response.success:
            return response.data
        raise ReadUnnamedError

    def read(self, size=None):
        buf = []
        # if size is unspecified, implementation-dependent
        # faasten now returns at most one block (4K) data
        if size is None:
            return self._blob_read()
        else:
            while size > 0:
                data = self._blob_read(size)
                # reaches EOF
                if len(data) == 0:
                    return buf
                buf.extend(data)
                offset += len(data)
                size -= len(data)
        # size = 0
        return buf

class CreateUnnamedError(Exception):
    pass

class ReadUnnamedError(Exception):
    pass

File: rootfs/runtimes/nodejs/syscalls.js
const vsock = require("vsock");
const path = require("path");
const syscalls_pb = require("./syscalls_pb")

class Syscall {
    constructor(sock) {
        this.sock = sock;
    }

    async _send(obj) {
        const objData = obj.serializeBinary(); // UInt8Array
        const buf = Buffer.alloc(4 + objData.length);
        let offset = buf.writeUInt32BE(objData.length);
        for (const b of objData) {
            offset = buf.writeUInt8(b, offset);
        }
        return await this.sock.write(buf);
    }

    async _recv(obj) {
        const lenBuf = await this.sock.read(4);
        const len = lenBuf.readUInt32BE();
        const objBuff = await this.sock.read(len);
        return obj
            .constructor
            .deserializeBinary(objBuff);
    }

    async request() {
        const request = new syscalls_pb.Request();
        return await this._recv(request);
    }

    async respond(res) {
        const _response = new syscalls_pb.Response();
        _response.setPayload(JSON.stringify(res));
        const response = new syscalls_pb.Syscall();
        response.setResponse(_response);
        await this._send(response);
    }

    async write_key(key, value) {
        const writeKey = new syscalls_pb.WriteKey();
        writeKey.setKey(key);
        writeKey.setValue(value);
        const req = new syscalls_pb.Syscall();
        req.setWritekey(writeKey);
        await this._send(req);
        const response =
            await this._recv(new syscalls_pb.WriteKeyResponse());
        return response.getSuccess();
    }

    async read_key(key) {
        const readKey = new syscalls_pb.ReadKey();
        readKey.setKey(key);
        const req = new syscalls_pb.Syscall();
        req.setReadkey(readKey);
        await this._send(req);
        const response =
            await this._recv(new syscalls_pb.ReadKeyResponse());
        return response.getValue();
    }

    async read_dir(d) {
        const readDir = new syscalls_pb.ReadDir();
        readDir.setDir(Buffer.from(d, "utf-8"));
        const req = new syscalls_pb.Syscall();
        req.setReaddir(readDir);
        await this._send(req);
        const response =
            await this._recv(new syscalls_pb.ReadDirResponse());
        const l = response.getKeysList();
        return l.map(b => Buffer.from(b).toString("utf-8"));
    }

    /* Label APIs */
    async get_current_label() {
        const label = new syscalls_pb.GetCurrentLabel();
        const req = new syscalls_pb.Syscall();
        req.setGetcurrentlabel(label);
        await this._send(req);
        const response =
            await this._recv(new syscalls_pb.DcLabel());
        return response;
    }

    async taint(label) {
        const req = new syscalls_pb.Syscall();
        req.setTaintwithlabel(label);
        await this._send(req)
        const response =
            await this._recv(new syscalls_pb.DcLabel());
        return response;
    }

    /**
     * @type {syscalls_pb.Component} secrecy
     */
    async declassify(secrecy) { /* TODO */ }
    /* End of Label APIs */

    /* Github APIs */
    async github_rest_get(route, token, toblob=false) {
        const rest = new syscalls_pb.GithubRest();
        rest.setVerb(syscalls_pb.HttpVerb.GET);
        rest.setRoute(route);
        rest.setBody(null);
        rest.setToblob(toblob);
        rest.setToken(token);
        const req = new syscalls_pb.Syscall();
        req.setGithubrest(rest);
        await this._send(req);
        const response =
            await this._recv(new syscalls_pb.GithubRestResponse());
        return response;
    }

    async github_rest_post(route, body, token, toblob=false) {
        const bodyJson = JSON.stringify(body);
        const rest = new syscalls_pb.GithubRest();
        rest.setVerb(syscalls_pb.HttpVerb.POST);
        rest.setRoute(route);
        rest.setBody(bodyJson);
        rest.setToblob(toblob);
        rest.setToken(token);
        const req = new syscalls_pb.Syscall();
        req.setGithubrest(rest);
        await this._send(req);
        const response =
            await this._recv(new syscalls_pb.GithubRestResponse());
        return response;
    }

    async github_rest_put(route, body, token, toblob=false) {
        const bodyJson = JSON.stringify(body);
        const rest = new syscalls_pb.GithubRest();
        rest.setVerb(syscalls_pb.HttpVerb.PUT);
        rest.setRoute(route);
        rest.setBody(bodyJson);
        rest.setToblob(toblob);
        rest.setToken(token);
        const req = new syscalls_pb.Syscall();
        req.setGithubrest(rest);
        await this._send(req);
        const response =
            await this._recv(new syscalls_pb.GithubRestResponse());
        return response;
    }

    async github_rest_delete(route, body, token, toblob=false) {
        const bodyJson = JSON.stringify(body);
        const rest = new syscalls_pb.GithubRest();
        rest.setVerb(syscalls_pb.HttpVerb.DELETE);
        rest.setRoute(route);
        rest.setBody(bodyJson);
        rest.setToblob(toblob);
        rest.setToken(token);
        const req = new syscalls_pb.Syscall();
        req.setGithubrest(rest);
        await this._send(req);
        const response =
            await this._recv(new syscalls_pb.GithubRestResponse());
        return response;
    }
    /* End of Github APIs */


    async invoke(f, payload) {
        const invoke = new syscalls_pb.Invoke();
        invoke.setFunction(f);
        invoke.setPayload(payload);
        const req = new syscalls_pb.Syscall();
        req.setInvoke(invoke);
        await this._send(req);
        const response =
            await this._recv(new syscalls_pb.InvokeResponse());
        return response.getSuccess();
    }

    async fs_read(p) {
        const fsRead = new syscalls_pb.FSRead();
        fsRead.setPath(p);
        const req = new syscalls_pb.Syscall();
        req.setFsread(fsRead);
        await this._send(req);
        const response =
            await this._recv(new syscalls_pb.ReadKeyResponse());
        return response.getValue();
    }

    async fs_write(p, data) {
        const fsWrite = new syscalls_pb.FSWrite();
        fsWrite.setPath(p);
        fsWrite.setData(data);
        const req = new syscalls_pb.Syscall();
        req.setFswrite(fsWrite);
        await this._send(req);
        const response =
            await this._recv(new syscalls_pb.WriteKeyResponse());
        return response.getSuccess();
    }

    /**
     * @type {syscalls_pb.DcLabel} label
     */
    async fs_createdir(p, label=null) {
        const dir = path.dirname(p);
        const name = path.basename(p);
        const fsDir = new syscalls_pb.FSCreateDir();
        fsDir.setBasedir(dir);
        fsDir.setName(name);
        fsDir.setLabel(label);
        const req = new syscalls_pb.Syscall();
        req.setFscreatedir(fsDir);
        await this._send(req);
        const response =
            await this._recv(new syscalls_pb.WriteKeyResponse())
        return response.getSuccess();
    }

    /**
     * @type {syscalls_pb.DcLabel} label
     */
    async fs_createfile(p, label=null) {
        const dir = path.dirname(p);
        const name = path.basename(p);
        const fsFile = new syscalls_pb.FSCreateFile();
        fsFile.setBasedir(dir);
        fsFile.setName(name);
        fsFile.setLabel(label);
        const req = new syscalls_pb.Syscall();
        req.setFscreatefile(fsFile);
        await this._send(req);
        const response =
            await this._recv(new syscalls_pb.WriteKeyResponse())
        return response.getSuccess();
    }


}
module.exports.Syscall = Syscall;


class NewBlob {
    constructor(fd, syscall) {
        this.fd = fd;
        this.syscall = syscall;
    }

    async write(data) {
        const blob = new syscalls_pb.BlobWrite();
        blob.setFd(this.fd);
        blob.setData(data);
        const req = new syscalls_pb.Syscall();
        req.setWriteblob(blob);
        await this.syscall._send(req);
        const response =
            await this.syscall._recv(new syscalls_pb.BlobResponse());
        return response.getSuccess();
    }

    async finalize(data) {
        const blob = new syscalls_pb.BlobFinalize();
        blob.setFd(this.fd);
        blob.setData(data);
        const req = new syscalls_pb.Syscall();
        req.setFinalizeblob(blob);
        await this.syscall._send(req);
        const response =
            await this.syscall._recv(new syscalls_pb.BlobResponse());
        const buf = Buffer.from(response.getData());
        return buf.toString("utf-8");
    }
}

class Blob {
    constructor(fd, syscall) {
        this.fd = fd;
        this.syscall = syscall;
    }

    async _blob_read(offset=null, length=null) {
        const blob = new syscalls_pb.BlobRead();
        blob.setFd(this.fd);
        blob.setOffset(offset);
        blob.setLength(length);
        const req = new syscalls_pb.Syscall();
        req.setReadblob(blob);
        await this.syscall._send(req);
        const response =
            await this.syscall._recv(new syscalls_pb.BlobResponse());
        if (response.getSuccess()) {
            return response.getData(); // TODO or getData_asU8?
        } else {
            // TODO error handling
        }
    }

    async read(size=null) {
        let buf = Buffer.alloc(0);
        if (size == null) {
            return await this._blob_read();
        } else {
            while (size > 0) {
                const data = await this._blob_read(size);
                // reaches EOF
                if (data.length == 0) {
                    return buf;
                }
                // assume data is an u8 array
                buf = Buffer.concat(buffer, data);
                offset += data.length;
                size -= data.length;
            }
        }
        return buf;
    }
}















